### PR TITLE
#160451497 send email to user on signup and verify user account

### PR DESCRIPTION
**Pull request by onwuvic**
#### What does this PR do?
It send user verification mail and verify user through the mail

#### Description of Task to be completed?
- Create a sendMail function to send mail to user using SendGrid
- Send verification mail to user upon signing up
- User can verify their account by clicking on the link in the mail
- Create a verifyUser method that verify user when the mail link is clicked
- Verify user and send a success message

#### How should this be manually tested?
- Clone the repo with "git clone https://github.com/andela/phlox-ah.git"
- cd into phlox-ah. cd phlox-ah
- Switch to the chore branch git branch -r then run git checkout feature/160451497/verification-email
- Run npm install
- Rename .env_sample file to .env then fill it with your locally created Postgres database details
- If you have PostgreSQL install you can create a database on your command line with createdb your_db_name
- Run npm run migrate-reset to migrate your table into your database
- Run npm run dev to start your server
- Open postman, then visit localhost:3000/api/v1/signup a Post request with the following form entries: email, username and password.
- A mail will be sent to your email account click to verify your account
- A successful registration completed will be sent to you

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
Story type: feature
Story title: An Email should be sent to a user upon successful registration. User should be able to receive a verification link via email upon successful signup
Story ID: #160451497

#### Screenshots (if appropriate)
N/A
#### Questions:```
N/A

### CODE REVIEW

File: server/controllers/userController.js [hunk 1]
```diff
@@ -20,13 +23,58 @@ export default class UserController {
   */
   static create(req, res) {
     const { username, email, password } = req.body;
+
+    // generate verification token and store in verifyToken variable
+    const verifyToken = crypto.randomBytes(16).toString('hex');
+
     bcrypt.hash(password, saltRounds)
-      .then(hashedPassword => User.create({ username, email, password: hashedPassword, })
```
Comment (onwuvic) [inline server/controllers/userController.js:26]: I couldn't put everything on one line. eslint flag it to be more than 100 characters in one line

File: server/tests/user.test.js [hunk 2]
```diff
@@ -149,7 +167,7 @@ describe('Users', () => {
         .end((err, res) => {
           expect(res.status).to.equal(201);
           expect(res.body).to.have.property('token');
-          expect(res.body.message).to.equal('You have successfully signed up!');
```
Comment (onwuvic) [inline server/tests/user.test.js]: User message upon registration was changed

**Comment by kampkelly** [comment]
LGTM

**Comment by udochukwu** [comment]
LGTM

**Comment by udochukwu** [comment]
LGTM

File: .env_sample [hunk 3]
```diff
@@ -3,6 +3,8 @@ DB_PASS=database_password
 DB_NAME_DEVELOPMENT=database_name
 DB_HOST=host
 DB_PORT=port
-DB_DIALECT=postgres,mysql,sqlite
```
Comment (onwuvic) [inline .env_sample]: No need to add mysql and sqlite

**Review by slimsolz — COMMENTED** [review]
LGTM

File: server/controllers/userController.js [hunk 4]
```diff
@@ -20,13 +23,58 @@ export default class UserController {
...
+      .then(hashedPassword => User.create({
+        username, email, password: hashedPassword, verifyToken
+      })
         .then((user) => {
           const { id } = user;
           const token = generateToken({ id });
-          res.status(201).json({ success: true, message: 'You have successfully signed up!', token });
```
Comment (onwuvic) [inline server/controllers/userController.js]: Change message to 'User successfully signed up, an email is sent to your mail account, please verify your mail account to complete registration' to inform the user to check the mail and verify their account.

File: server/middlewares/authenticator.js [hunk 5]
```diff
@@ -10,7 +10,9 @@ export default {
   },
 
   checkToken(req, res, next) {
-    const token = req.body.token || req.headers['x-access-token'] || req.headers.bearer;
```
Comment (onwuvic) [inline server/middlewares/authenticator.js]: There is no for req.body.token and req.headers.bearer does not exist

**Comment by samdiano** [comment]
LGTM

**Review by s-akinrele — COMMENTED** [review]
GTM .. But confirm with the PO for copy text

**Review by s-akinrele — APPROVED** [review]
GTM

**Pull request merged**